clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -1611,4 +1611,208 @@ mlir::Value ScalarExprEmitter::VisitAbstractConditionalOperator(
+              SmallVector<mlir::Location, 2> Locs;
+              if (Loc.isa<mlir::FileLineColLoc>()) {
+                Locs.push_back(Loc);
+                Locs.push_back(Loc);
+              } else if (Loc.isa<mlir::FusedLoc>()) {
+                auto fusedLoc = Loc.cast<mlir::FusedLoc>();
+                Locs.push_back(fusedLoc.getLocations()[0]);
+                Locs.push_back(fusedLoc.getLocations()[1]);
+              }
+              CIRGenFunction::LexicalScopeContext lexScope{
+                  Locs[0], Locs[1], B.getInsertionBlock()};
+              CIRGenFunction::LexicalScopeGuard lexElseGuard{CGF, &lexScope};
+              CGF.currLexScope->setAsTernary();
+              auto res = B.create<mlir::cir::ConstantOp>(
+                  Loc, Builder.getBoolTy(),
+                  mlir::BoolAttr::get(B.getContext(), true));
+              B.create<mlir::cir::YieldOp>(Loc, res.getRes());
+            },
+            /*falseBuilder*/
+            [&](mlir::OpBuilder &b, mlir::Location Loc) {
+              SmallVector<mlir::Location, 2> Locs;
+              if (Loc.isa<mlir::FileLineColLoc>()) {
+                Locs.push_back(Loc);
+                Locs.push_back(Loc);
+              } else if (Loc.isa<mlir::FusedLoc>()) {
+                auto fusedLoc = Loc.cast<mlir::FusedLoc>();
+                Locs.push_back(fusedLoc.getLocations()[0]);
+                Locs.push_back(fusedLoc.getLocations()[1]);
+              }
+              CIRGenFunction::LexicalScopeContext lexScope{
+                  Locs[0], Locs[1], b.getInsertionBlock()};
+              CIRGenFunction::LexicalScopeGuard lexElseGuard{CGF, &lexScope};
+              CGF.currLexScope->setAsTernary();
+              auto res = b.create<mlir::cir::ConstantOp>(
+                  Loc, Builder.getBoolTy(),
+                  mlir::BoolAttr::get(b.getContext(), false));
+              b.create<mlir::cir::YieldOp>(Loc, res.getRes());
+            });
+        B.create<mlir::cir::YieldOp>(Loc, res.getResult());
+      },
+      /*falseBuilder*/
+      [&](mlir::OpBuilder &B, mlir::Location Loc) {
+        SmallVector<mlir::Location, 2> Locs;
+        if (Loc.isa<mlir::FileLineColLoc>()) {
+          Locs.push_back(Loc);
+          Locs.push_back(Loc);
+        } else if (Loc.isa<mlir::FusedLoc>()) {
+          auto fusedLoc = Loc.cast<mlir::FusedLoc>();
+          Locs.push_back(fusedLoc.getLocations()[0]);
+          Locs.push_back(fusedLoc.getLocations()[1]);
+        }
+        CIRGenFunction::LexicalScopeContext lexScope{Loc, Loc,
+                                                     B.getInsertionBlock()};
+        CIRGenFunction::LexicalScopeGuard lexElseGuard{CGF, &lexScope};
+        CGF.currLexScope->setAsTernary();
+        auto res = B.create<mlir::cir::ConstantOp>(
+            Loc, Builder.getBoolTy(),
+            mlir::BoolAttr::get(B.getContext(), false));
+        B.create<mlir::cir::YieldOp>(Loc, res.getRes());
+      });
```

Comment:
Looks like your ending this function assuming the result is always `cir.bool`, which might not be the case. For instance, note how LLVM's implementation use `return Builder.CreateZExtOrBitCast(PN, ResTy, "land.ext");`. If you only wanna handle bools here, please add an unreacheable to assert on unhandled cases.